### ISSUE TITLE
Include aar packaging type for cwac-sacklist dependency in pom.

### ISSUE THREAD

**Issue by mhelder** [commonsguy/cwac-merge#19]
Using maven to add cwac-merge as a dependency to an Android app, the following error is thrown when resolving the newly added dependency:

> [ERROR] Failed to execute goal on project [project_name]: Could not resolve dependencies for project [app_name]: Could not find artifact com.commonsware.cwac:sacklist:jar:1.0.+ in commonsware (http://repo.commonsware.com.s3.amazonaws.com) -> [Help 1]

The problem here is that [merge-1.1.1.pom](http://repo.commonsware.com.s3.amazonaws.com/com/commonsware/cwac/merge/1.1.1/merge-1.1.1.pom) points to cwac-sacklist as a dependency, but doesn't define the packaging type. As such, maven defaults to looking for a jar file (as illustrated by the error message above). But, as we can see in the [repo](http://repo.commonsware.com.s3.amazonaws.com/), sacklist is only available as aar package.

Specifying the packaging type (`aar`) for the nested cwac-sacklist dependency solves the issue. That is, with the pom file below, the nested dependency can be resolved correctly:

    <?xml version="1.0" encoding="UTF-8"?>
    <project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <modelVersion>4.0.0</modelVersion>
      <groupId>com.commonsware.cwac</groupId>
      <artifactId>merge</artifactId>
      <version>1.1.1</version>
      <packaging>aar</packaging>
      <dependencies>
        <dependency>
          <groupId>com.commonsware.cwac</groupId>
          <artifactId>sacklist</artifactId>
          <version>1.0.+</version>
          <type>aar</type>
          <scope>compile</scope>
        </dependency>
      </dependencies>
    </project>

(Note: only change is the added line that says: `<type>aar</type>`)

It'd be great if the pom could be updated to include the aar packaging type for the sacklist dependency, allowing maven to correctly resolve it (or alternatively, publish a `jar` package too). I haven't tested with Gradle (yet), but I suspect it may default to `aar` packaging?

**Comment by commonsguy** [comment]
@mhelder I do not support Maven builds, just as I do not support builds using Buck, etc. I support Gradle builds and classic Eclipse/Ant builds. While I don't mind it if Maven works, it is not an objective, and I do not want to harm my project in the interests of adding Maven support.

> I haven't tested with Gradle (yet), but I suspect it may default to aar packaging?

AFAIK, it looks for either, preferring AAR when that is found. IMHO, Maven's Android support needs to look for AAR and JAR artifacts. Quoting [Jake Wharton](https://groups.google.com/d/msg/adt-dev/H-tB6SJFh5Y/SeJ2ZTr6W50J): "You almost never want to specify explicit packaging" in Gradle builds.

And, of course, in the fullness of time, `cwac-merge` and `cwac-sacklist` are not important, as they support `ListView` and `GridView`, and those will be replaced by `RecyclerView`. 

Is there some way, in your app's Maven build script, to direct Maven to use AARs for the `cwac-sacklist` packaging?

**Comment by mhelder** [comment]
@commonsguy: Fair enough, and thanks for the quick reply. 

> IMHO, Maven's Android support needs to look for AAR and JAR artifacts.

I agree. Then I wouldn't have had this problem in the first place. Unfortunately, that doesn't appear to be the case. 

> Is there some way, in your app's Maven build script, to direct Maven to use AARs for the cwac-sacklist packaging?

There is, but only for direct project dependencies. Any transitive/nested dependencies are resolved using the dependency's own pom file, which leads back to the issue outlined described in my earlier post. I've Googled the crap out of it, but as far as I can tell there is no way to instruct maven to pick up `aar` packages too as an 'alternative default'. Basically, this same issue is pointed out by W. Ferguson in the thread you link to (scroll down to the last reply).

Eventually the app I'm working on is going to migrate to `RecyclerView`, but in order to retain the same behaviour it's going to be a non-trivial change. A fairly complex maven build pipeline currently set up for this project is also holding us back from moving to Gradle in the near future.

For those Maven users out there, would it be an option to also publish a `jar` package of `cwac-merge` and/or `cwac-sacklist` in the repository? That shouldn't harm your projects or supported build systems in any way. As a matter of fact, you already offer `jar` packages as standalone download - publishing them to the repo would help us out greatly.

**Comment by commonsguy** [comment]
@mhelder 

> Basically, this same issue is pointed out by W. Ferguson in the thread you link to (scroll down to the last reply).

Yeah, he posted that after I responded to your issue, IIRC.

> That shouldn't harm your projects or supported build systems in any way.

My vision is to only have AARs in the repo for all of the CWAC libraries. Primarily, that is because I do not want to support JARs for the long term, as that locks me into not having resources, manifest entries, etc. I publish JARs as a stopgap, particularly for libraries that predated AARs, in support of Eclipse users, but eventually I plan to do away with the JARs altogether.

My recommendation, at this point, is for you to switch to source. The `MergeAdapter` class is only a bit longer than is this issue, and `SackOfViewsAdapter` is shorter yet. I am not planning any changes to these classes, and so you are not at significant risk of "falling behind" what's published in the AARs. And that way, the source code can be built by whatever means you want (in this case, Maven).

**Comment by mhelder** [comment]
@commonsguy 

> My recommendation, at this point, is for you to switch to source.

Yeah, that's what I decided to do in the first place. :) It beats the alternatives of manually editing our local Maven repo's pom file and/or bundling the individual `jar` packages as part of the project. But just by a little bit... ;) (if anyone who stumbles on this is aware of any other solutions: please do tell!)

The mid- to long-term goals are to both migrate to Gradle as well as use `RecyclerView` as replacement for the two `cwac` dependencies. As soon as one or both happen, the current issue will be resolved. As such, I'll close this ticket for now.

Thanks again for your feedback, it's much appreciated.

**Comment by mhelder** [comment]
> (if anyone who stumbles on this is aware of any other solutions: please do tell!)

I'll tell myself. :) I just realised a much cleaner approach to work around the issue described above exists. Not sure why I didn't think of it before, especially since it's so obvious, but here goes.

    <dependency>
        <groupId>com.commonsware.cwac</groupId>
        <artifactId>merge</artifactId>
        <version>1.1.1</version>
        <type>aar</type>
        <exclusions>
            <exclusion>
                <groupId>com.commonsware.cwac</groupId>
                <artifactId>sacklist</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
    <dependency>
        <groupId>com.commonsware.cwac</groupId>
        <artifactId>sacklist</artifactId>
        <version>1.0.2</version>
        <type>aar</type>
    </dependency>

So, that's all there's to it: I'm explicitly skipping the problematic (for Maven) transitive dependency `cwac-sacklist` and plugging the `aar` package as a direct dependency into the project. This Maven doesn't go looking for a (non-existent) `jar`, and we still get all the benefits of maven-hosted artifacts.

For those interested: I also played around with adding the two dependencies as `jar` packages to the project, by means of including them as `system` packages, but ran into all sorts of issues with Proguard and integration tests afterwards. While I managed to overcome the initial Proguard issues, for some reason the integration tests were still throwing `ClassNotFoundException`s on `MergeAdapter`, even though the app itself worked fine. 

Oh well... sometimes simple solutions can be best.
